src/main/java/com/minecolonies/coremod/colony/ColonyList.java

```diff
@@ -0,0 +1,240 @@
+import java.util.NoSuchElementException;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+//TODO extend list/collection
+/**
+ * Data structure for storing colonies, optimized for performance.
+ *
+ * @author Colton
+ */
+public final class ColonyList<T extends IColony> implements Iterable<T>
+{
+    private static final int INITIAL_SIZE = 16;
+
+    private Object[] list = new Object[INITIAL_SIZE];
+
+    private final List<Integer> nullIndices = new ArrayList<>();
+
+    private int topID = 0;
+
+    private int size = 0;
+
+    /**
+     * Get the Colony with the provided colony id.
+     * @param index colony id.
+     * @return The Colony associated with the provided id.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public T get(int index)
+    {
+        if (index < 1 || index > list.length)
+        {
+            return null;
+        }
+
+        return (T) list[index];
+    }
+
+    /**
+     * Creates a new Colony, adds it to the list, and returns it.
+     * @param world The world for the Colony.
+     * @param position The position for the Colony center.
+     * @return The newly created Colony.
+     */
+    @NotNull
+    public Colony create(World world, BlockPos position)
+    {
+        final Colony colony = new Colony(getNextColonyID(), world, position);
+        list[colony.getID()] = colony;
+        size++;
+        return colony;
+    }
+
+    /**
+     * Add a new Colony to the List.
+     * @param colony colony to add to the list.
+     */
+    public void add(T colony)
+    {
+        final T existingColony = get(colony.getID());
+        if (existingColony != null && existingColony != colony)
+        {
+            throw new IllegalArgumentException(
+                    String.format("Already a colony registered to id=%d, colony=%s, not changing to colony=%s",
+                            colony.getID(),
+                            existingColony.getName(),
+                            colony.getName()));
+        }
+
+        list[colony.getID()] = colony;
+    }
+
+    /**
+     * Remove the Colony from the list.
+     * @param colony the Colony to remove.
+     */
+    public void remove(T colony)
+    {
+        remove(colony.getID());
+    }
+
+    /**
+     * Remove the colony with the provided id from the list.
+     * @param id colony id to remove.
+     */
+    public void remove(int id)
+    {
+        if (list[id] == null)
+        {
+            Log.getLogger().warn("Tried to remove colony with id=%d, but it didn't exist.", id);
+        }
+
+        size--;
+        list[id] = null;
+
+        if (!nullIndices.contains(id))
+        {
+            nullIndices.add(id);
+        }
+    }
+
+    private int getNextColonyID()
+    {
+        if (nullIndices.isEmpty())
+        {
+            topID++;
+            if (topID >= list.length)
+            {
+                // Expand list
+                Object[] newList = new Object[list.length * 2];
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'newList' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)